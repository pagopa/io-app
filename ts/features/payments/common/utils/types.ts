export type AlertVariant = "error" | "success" | "warning" | "info";

export type PaymentsBackoffRetryValue = {
  retryCount: number;
  allowedRetryTimestamp?: number;
};
