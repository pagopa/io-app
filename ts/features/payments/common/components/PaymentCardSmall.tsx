import {
  Chip,
  IOColors,
  IOStyles,
  Icon,
  VSpacer
} from "@pagopa/io-app-design-system";
import * as React from "react";
import { StyleSheet, View } from "react-native";
import Placeholder from "rn-placeholder";
import { LogoPaymentWithFallback } from "../../../../components/ui/utils/components/LogoPaymentWithFallback";
import { WithTestID } from "../../../../types/WithTestID";
import { PaymentCardProps } from "./PaymentCard";
import { PaymentCardPressableBase } from "./PaymentCardPressableBase";

export type PaymentCardSmallProps = WithTestID<
  PaymentCardProps & {
    bankName?: string;
    onPress?: () => void;
    isError?: boolean;
    accessibilityLabel?: string;
  }
>;

const PaymentCardSmall = ({
  testID,
  onPress,
  isError,
  accessibilityLabel,
  ...props
}: PaymentCardSmallProps) => {
  const labelText = React.useMemo(() => {
    if (props.hpan) {
      return `•••• ${props.hpan}`;
    }

    if (props.holderEmail) {
      return "PayPal";
    }

    if (props.holderPhone) {
      return "BANCOMAT Pay";
    }

    if (props.bankName) {
      return props.bankName;
    }

    return props.brand;
  }, [props]);

  const iconName = React.useMemo(() => {
    if (props.holderEmail) {
      return "paypal";
    }

    if (props.holderPhone) {
      return "bancomatpay";
    }

    return props.brand;
  }, [props]);

  return (
    <PaymentCardPressableBase
      onPress={onPress}
      testID={`${testID}-pressable`}
      accessibilityLabel={accessibilityLabel}
    >
      <View style={[styles.card, isError && styles.cardError]} testID={testID}>
        <View style={[IOStyles.rowSpaceBetween, IOStyles.alignCenter]}>
          <LogoPaymentWithFallback brand={iconName} size={24} />
          {isError && (
            <Icon
              testID={`${testID}-errorIcon`}
              name="errorFilled"
              size={16}
              color="error-850"
            />
          )}
        </View>
        <VSpacer size={8} />
        <Chip
          ellipsizeMode="tail"
          weight="Regular"
          numberOfLines={1}
          color={isError ? "error-850" : "grey-700"}
        >
          {labelText}
        </Chip>
      </View>
    </PaymentCardPressableBase>
  );
};

const PaymentCardSmallSkeleton = ({ testID }: WithTestID<unknown>) => (
  <View style={styles.card} testID={`${testID}-skeleton`}>
    <Placeholder.Box
      color={IOColors["grey-200"]}
      animate="fade"
      radius={12}
      width={24}
      height={24}
    />
    <VSpacer size={8} />
    <Placeholder.Box
      color={IOColors["grey-200"]}
      animate="fade"
      radius={8}
      width={"100%"}
      height={16}
    />
  </View>
);

export const PAYMENT_CARD_SMALL_WIDTH = 127;
export const PAYMENT_CARD_SMALL_HEIGHT = 80;

const styles = StyleSheet.create({
  card: {
    flex: 1,
    width: PAYMENT_CARD_SMALL_WIDTH,
    height: PAYMENT_CARD_SMALL_HEIGHT,
    flexBasis: PAYMENT_CARD_SMALL_WIDTH,
    flexGrow: 0,
    borderRadius: 8,
    padding: 16,
    aspectRatio: 16 / 10,
    backgroundColor: IOColors["grey-100"]
  },
  cardError: {
    backgroundColor: IOColors["error-100"]
  }
});

export { PaymentCardSmall, PaymentCardSmallSkeleton };
