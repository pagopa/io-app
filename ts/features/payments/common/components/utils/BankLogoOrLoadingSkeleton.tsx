import { IOColors, IOSkeleton } from "@pagopa/io-app-design-system";
import { useEffect, useState } from "react";
import { Image } from "react-native";
import { getBankLogosCdnUri } from "../../../../../components/ui/utils/strings";

type BankLogoOrSkeletonProps = {
  abiCode?: string;
  dimensions: { height: number; width: number };
  placeHolderColor?: IOColors;
  imageA11yLabel?: string;
};

/**
 *component that renders a bank logo based on the abiCode,
 * and displays a loader while the image is loading
 * @param abiCode
 * @param dimensions - height and width of the image and placeholder
 * @param placeHolderColor - color of the placeholder
 */
export const BankLogoOrSkeleton = ({
  abiCode,
  dimensions,
  placeHolderColor = "grey-200",
  imageA11yLabel
}: BankLogoOrSkeletonProps) => {
  const [imageUrl, setImageUrl] = useState<string | undefined>(undefined);
  const { height, width: maxWidth } = dimensions;
  const [width, setWidth] = useState<number>(maxWidth);
  useEffect(() => {
    // we pre-fetch the image to avoid having to render both items
    // at the same time, which looks like an untidy hack
    if (abiCode === undefined) {
      return;
    }
    fetchBlob(getBankLogosCdnUri(abiCode))
      .then(blob => {
        const url = URL.createObjectURL(blob);
        // to make sure the ratio is correct, we
        // calculate it ourselves, else we risk misalignments
        Image.getSize(url, (width, imgHeight) => {
          const ratio = width / imgHeight;
          setWidth(ratio * height);
        });
        setImageUrl(url);
      })
      .catch(_ => null);
  }, [abiCode, height]);

  return imageUrl !== undefined ? (
    <Image
      accessibilityIgnoresInvertColors
      accessible
      accessibilityLabel={imageA11yLabel}
      source={{ uri: imageUrl }}
      style={{
        height,
        width,
        resizeMode: "contain"
      }}
    />
  ) : (
    <IOSkeleton
      color={IOColors[placeHolderColor]}
      shape="rectangle"
      radius={8}
      height={height}
      width={width}
    />
  );
};
async function fetchBlob(url: string) {
  const response = await fetch(url);
  if (response.status === 200) {
    return response.blob();
  } else {
    throw new Error();
    // error is caught and irrelevant
  }
}
