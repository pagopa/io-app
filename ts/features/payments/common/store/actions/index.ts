import { PaymentsMethodDetailsActions } from "../../../details/store/actions";
import { PaymentsHistoryActions } from "../../../history/store/actions";
import { PaymentsOnboardingActions } from "../../../onboarding/store/actions";
import { PaymentsCheckoutActions } from "../../../checkout/store/actions";
import { PaymentsTransactionActions } from "../../../transaction/store/actions";
import { PaymentsWalletActions } from "../../../wallet/store/actions";

export type PaymentsActions =
  | PaymentsOnboardingActions
  | PaymentsMethodDetailsActions
  | PaymentsCheckoutActions
  | PaymentsTransactionActions
  | PaymentsHistoryActions
  | PaymentsWalletActions;
