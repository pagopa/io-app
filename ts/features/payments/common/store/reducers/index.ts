import { combineReducers } from "redux";
import { PersistPartial } from "redux-persist";
import detailsReducer, {
  PaymentsMethodDetailsState
} from "../../../details/store/reducers";
import historyReducer, {
  PaymentsHistoryState
} from "../../../history/store/reducers";
import onboardingReducer, {
  PaymentsOnboardingState
} from "../../../onboarding/store/reducers";
import paymentReducer, {
  PaymentsCheckoutState
} from "../../../checkout/store/reducers";
import transactionReducer, {
  PaymentsTransactionState
} from "../../../transaction/store/reducers";

export type PaymentsState = {
  onboarding: PaymentsOnboardingState;
  details: PaymentsMethodDetailsState;
  payment: PaymentsCheckoutState;
  transaction: PaymentsTransactionState;
  history: PaymentsHistoryState & PersistPartial;
};

const paymentsReducer = combineReducers({
  onboarding: onboardingReducer,
  details: detailsReducer,
  payment: paymentReducer,
  transaction: transactionReducer,
  history: historyReducer
});

export default paymentsReducer;
