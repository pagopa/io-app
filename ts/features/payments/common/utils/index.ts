import {
  IOLogoPaymentType,
  IOPaymentLogos,
  ListItemTransactionStatusWithBadge
} from "@pagopa/io-app-design-system";
import * as O from "fp-ts/lib/Option";
import { pipe } from "fp-ts/lib/function";
import I18n from "i18n-js";
import _ from "lodash";
import { Bundle } from "../../../../../definitions/pagopa/ecommerce/Bundle";
import { WalletApplicationStatusEnum } from "../../../../../definitions/pagopa/walletv3/WalletApplicationStatus";
import { WalletInfo } from "../../../../../definitions/pagopa/walletv3/WalletInfo";
import { PaymentSupportStatus } from "../../../../types/paymentMethodCapabilities";
import { getDateFromExpiryDate, isExpiredDate } from "../../../../utils/dates";
import { WalletPaymentPspSortType } from "../../checkout/types";
import { PaymentCardProps } from "../components/PaymentCard";
import { UIWalletInfoDetails } from "../types/UIWalletInfoDetails";
import { findFirstCaseInsensitive } from "../../../../utils/object";
import { WalletCard } from "../../../newWallet/types";

/**
 * A simple function to get the corresponding translated badge text,
 * based on the transaction status.
 */

export const getBadgeTextByTransactionStatus = (
  transactionStatus: ListItemTransactionStatusWithBadge
) => {
  switch (transactionStatus) {
    case "failure":
      return I18n.t("global.badges.failed");
    case "cancelled":
      return I18n.t("global.badges.cancelled");
    case "reversal":
      return I18n.t("global.badges.reversal");
    case "pending":
      return I18n.t("global.badges.onGoing");
    default:
      return "";
  }
};

/**
 * Check if the given payment method is expired
 * right(true) if it is expired, right(false) if it is still valid
 * left if expiring date can't be evaluated
 * @param paymentMethod
 */
export const isPaymentMethodExpired = (
  details?: UIWalletInfoDetails
): boolean =>
  pipe(
    details?.expiryDate,
    O.fromNullable,
    O.chainNullableK(getDateFromExpiryDate),
    O.chainNullableK(isExpiredDate),
    O.getOrElse(() => false)
  );

/**
 * true if the given paymentMethod supports the given walletFunction
 * @param paymentMethod
 * @param walletFunction
 */
export const hasApplicationEnabled = (
  paymentMethod: WalletInfo | undefined,
  walletApplication: string
): boolean =>
  paymentMethod !== undefined &&
  paymentMethod.applications.some(
    application =>
      application.name === walletApplication &&
      application.status === WalletApplicationStatusEnum.ENABLED
  );
/**
 * return true if the payment method has the payment feature
 */
export const hasPaymentFeature = (paymentMethod: WalletInfo): boolean =>
  hasApplicationEnabled(paymentMethod, "PAGOPA");

/**
 * Check if a payment method is supported or not
 * If the payment method have the enableable function pagoPA, can always pay ("available")
 * "available" -> can pay
 * "arriving" -> will pay
 * "notAvailable" -> can't pay
 * "onboardableNotImplemented" -> can onboard a card that can pay but is not yet implemented
 */
export const isPaymentSupported = (
  paymentMethod: WalletInfo
): PaymentSupportStatus => {
  const paymentSupported: O.Option<PaymentSupportStatus> = hasPaymentFeature(
    paymentMethod
  )
    ? O.some("available")
    : O.none;

  const notAvailableCustomRepresentation = O.some(
    "notAvailable" as PaymentSupportStatus
  );

  return pipe(
    paymentSupported,
    O.alt(() => notAvailableCustomRepresentation),
    O.getOrElseW(() => "notAvailable" as const)
  );
};

export const WALLET_PAYMENT_TERMS_AND_CONDITIONS_URL =
  "https://www.pagopa.gov.it/it/prestatori-servizi-di-pagamento/elenco-PSP-attivi/";

/**
 * Function that returns a sorted list of psp based on the given sortType
 * The sortType can be: "name", "amount" or "default"
 */
export const getSortedPspList = (
  pspList: ReadonlyArray<Bundle>,
  sortType: WalletPaymentPspSortType
) => {
  switch (sortType) {
    case "name":
      return _.orderBy(pspList, psp => psp.bundleName);
    case "amount":
      return _.orderBy(pspList, psp => psp.taxPayerFee);
    case "default":
    default:
      return _.orderBy(pspList, ["onUs", "taxPayerFee"]);
  }
};

export const getPaymentCardPropsFromWalletInfo = (
  wallet: WalletInfo
): PaymentCardProps => {
  const details = wallet.details as UIWalletInfoDetails;
  const isExpired = isPaymentMethodExpired(details);

  return {
    hpan: details.lastFourDigits,
    brand: details.brand,
    expireDate: getDateFromExpiryDate(details.expiryDate),
    holderEmail: details.maskedEmail,
    holderPhone: details.maskedNumber,
    isExpired
  };
};

export const getPaymentLogoFromWalletDetails = (
  details: UIWalletInfoDetails
): IOLogoPaymentType | undefined => {
  if (details.maskedEmail !== undefined) {
    return "payPal";
  } else if (details.maskedNumber !== undefined) {
    return "bancomatPay";
  } else {
    return pipe(
      details.brand,
      O.fromNullable,
      O.chain(findFirstCaseInsensitive(IOPaymentLogos)),
      O.map(([logoName, _]) => logoName as IOLogoPaymentType),
      O.toUndefined
    );
  }
};

export const mapWalletIdToCardKey = (walletId: string) => `method_${walletId}`;

export const mapWalletsToCards = (
  wallets: ReadonlyArray<WalletInfo>
): ReadonlyArray<WalletCard> =>
  wallets.map<WalletCard>(wallet => ({
    ...getPaymentCardPropsFromWalletInfo(wallet),
    key: mapWalletIdToCardKey(wallet.walletId),
    type: "payment",
    category: "payment",
    walletId: wallet.walletId
  }));

/**
 * Function that returns a formatted payment notice number
 * by placing two spaces between every four numbers
 */
export const formatPaymentNoticeNumber = (noticeNumber: string) =>
  noticeNumber.replace(/(\d{4})/g, "$1  ").trim();
