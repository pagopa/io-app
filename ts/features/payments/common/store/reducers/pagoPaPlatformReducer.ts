import * as pot from "@pagopa/ts-commons/lib/pot";
import { getType } from "typesafe-actions";
import { Action } from "../../../../../store/actions/types";
import { NetworkError } from "../../../../../utils/errors";
import {
  clearPaymentsPendingActions,
  paymentsGetPagoPaPlatformSessionTokenAction,
  paymentsResetPagoPaPlatformSessionTokenAction,
  savePaymentsPendingAction
} from "../actions";

export type PaymentsPagoPaPlatformState = {
  sessionToken: pot.Pot<string, NetworkError>;
  pendingActions: Array<Action>;
};

const INITIAL_STATE: PaymentsPagoPaPlatformState = {
  sessionToken: pot.none,
  pendingActions: []
};

const reducer = (
  state: PaymentsPagoPaPlatformState = INITIAL_STATE,
  action: Action
): PaymentsPagoPaPlatformState => {
  switch (action.type) {
    case getType(paymentsGetPagoPaPlatformSessionTokenAction.request):
      return {
        ...state,
        sessionToken: pot.toLoading(state.sessionToken)
      };
    case getType(paymentsGetPagoPaPlatformSessionTokenAction.success):
      return {
        ...state,
        sessionToken: pot.some(action.payload.token as string)
      };
    case getType(paymentsGetPagoPaPlatformSessionTokenAction.failure):
      return {
        ...state,
        sessionToken: pot.toError(state.sessionToken, action.payload)
      };
    case getType(paymentsResetPagoPaPlatformSessionTokenAction):
      return {
        ...state,
        sessionToken: pot.none
      };
    // ADD PENDING ACTION WHILE REFRESHING THE PAGOPA TOKEN
    case getType(savePaymentsPendingAction):
      const actionToSave = action as ReturnType<
        typeof savePaymentsPendingAction
      >;
      return {
        ...state,
        pendingActions: [
          ...state.pendingActions,
          actionToSave.payload.pendingAction
        ]
      };
    case getType(clearPaymentsPendingActions):
      return {
        ...state,
        pendingActions: []
      };
  }
  return state;
};

export default reducer;
