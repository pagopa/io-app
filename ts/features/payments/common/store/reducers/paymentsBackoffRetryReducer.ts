import { getType } from "typesafe-actions";
import { Action } from "../../../../../store/actions/types";
import {
  clearPaymentsBackoffRetry,
  increasePaymentsBackoffRetry
} from "../actions";
import {
  PAYMENTS_BACKOFF_SECONDS_DELAYS,
  SECONDS_TO_MILLISECONDS
} from "../../utils/backoffRetry";
import { PaymentsBackoffRetryValue } from "../../utils/types";

export type PaymentsBackoffRetryState = {
  [key: string]: PaymentsBackoffRetryValue | undefined;
};

const INITIAL_STATE: PaymentsBackoffRetryState = {};

const reducer = (
  state: PaymentsBackoffRetryState = INITIAL_STATE,
  action: Action
): PaymentsBackoffRetryState => {
  switch (action.type) {
    case getType(increasePaymentsBackoffRetry):
      const currentRetryCount = state[action.payload]?.retryCount ?? 0;
      const retryCount = currentRetryCount + 1;
      const allowedRetryTimestamp =
        Date.now() +
        PAYMENTS_BACKOFF_SECONDS_DELAYS[retryCount] * SECONDS_TO_MILLISECONDS;

      return {
        ...state,
        [action.payload]: {
          retryCount,
          allowedRetryTimestamp
        }
      };
    case getType(clearPaymentsBackoffRetry):
      if (!state[action.payload]) {
        return state;
      }
      return {
        ...state,
        [action.payload]: undefined
      };
  }
  return state;
};

export default reducer;
