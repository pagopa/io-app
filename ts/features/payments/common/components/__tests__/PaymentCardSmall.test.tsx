import { default as React } from "react";
import { createStore } from "redux";
import ROUTES from "../../../../../navigation/routes";
import { applicationChangeState } from "../../../../../store/actions/application";
import { appReducer } from "../../../../../store/reducers";
import { GlobalState } from "../../../../../store/reducers/types";
import { renderScreenWithNavigationStoreContext } from "../../../../../utils/testWrapper";
import { PaymentCardSmall, PaymentCardSmallProps } from "../PaymentCardSmall";

describe("PaymentCardSmall", () => {
  const testID = "PaymentCardSmallTestID";
  jest.useFakeTimers();

  it(`should match the snapshot`, () => {
    const component = renderCard({
      hpan: "9900",
      brand: "maestro",
      onPress: () => undefined,
      testID
    });
    expect(component).toMatchSnapshot();
  });

  it(`should render card without pressable wrapper`, () => {
    const { queryByTestId } = renderCard({
      testID
    });
    expect(queryByTestId(`${testID}-pressable`)).toBeNull();
  });

  it(`should render credit card details`, () => {
    const { queryByText, queryByTestId } = renderCard({
      hpan: "9900",
      brand: "maestro",
      onPress: () => undefined,
      testID
    });
    expect(queryByText("•••• 9900")).not.toBeNull();
    expect(queryByTestId(`${testID}-errorIcon`)).toBeNull();
    expect(queryByTestId(`${testID}-pressable`)).not.toBeNull();
  });

  it(`should render error card`, () => {
    const { queryByText, queryByTestId } = renderCard({
      expireDate: new Date(2023),
      isExpired: true,
      hpan: "9900",
      brand: "maestro",
      onPress: () => undefined,
      testID
    });
    expect(queryByText("•••• 9900")).not.toBeNull();
    expect(queryByTestId(`${testID}-errorIcon`)).not.toBeNull();
    expect(queryByTestId(`${testID}-pressable`)).not.toBeNull();
  });

  it(`should render paypal card details`, () => {
    const { queryByText, queryByTestId } = renderCard({
      holderEmail: "test@test.it",
      onPress: () => undefined,
      testID
    });
    expect(queryByText("PayPal")).not.toBeNull();
    expect(queryByTestId(`${testID}-pressable`)).not.toBeNull();
  });

  it(`should render bpay card details`, () => {
    const { queryByText, queryByTestId } = renderCard({
      holderPhone: "1234",
      onPress: () => undefined,
      testID
    });
    expect(queryByText("BANCOMAT Pay")).not.toBeNull();
    expect(queryByTestId(`${testID}-pressable`)).not.toBeNull();
  });
});

function renderCard(props: PaymentCardSmallProps) {
  const globalState = appReducer(undefined, applicationChangeState("active"));
  return renderScreenWithNavigationStoreContext<GlobalState>(
    () => <PaymentCardSmall {...props} />,
    ROUTES.WALLET_HOME,
    {},
    createStore(appReducer, globalState as any)
  );
}
