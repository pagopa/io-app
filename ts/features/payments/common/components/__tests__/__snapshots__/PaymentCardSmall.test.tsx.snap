// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaymentCardSmall should match the snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          Object {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "#FFFFFF",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": Object {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              Object {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                Object {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    Object {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  WALLET_HOME
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  Object {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          activityState={2}
          collapsable={false}
          forwardedRef={[Function]}
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              Array [
                Object {
                  "display": "flex",
                  "overflow": undefined,
                },
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                forwardedRef={[Function]}
                handlerTag={1}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "#FFFFFF",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        accessible={true}
                        collapsable={false}
                        focusable={true}
                        onBlur={[Function]}
                        onClick={[Function]}
                        onFocus={[Function]}
                        onResponderGrant={[Function]}
                        onResponderMove={[Function]}
                        onResponderRelease={[Function]}
                        onResponderTerminate={[Function]}
                        onResponderTerminationRequest={[Function]}
                        onStartShouldSetResponder={[Function]}
                        onTouchEnd={[Function]}
                        testID="PaymentCardSmallTestID-pressable"
                      >
                        <View
                          style={
                            Object {
                              "transform": Array [
                                Object {
                                  "scale": undefined,
                                },
                              ],
                            }
                          }
                        >
                          <View
                            style={
                              Array [
                                Object {
                                  "aspectRatio": 1.6,
                                  "backgroundColor": "#E8EBF1",
                                  "borderRadius": 8,
                                  "flex": 1,
                                  "flexBasis": 127,
                                  "flexGrow": 0,
                                  "height": 80,
                                  "padding": 16,
                                  "width": 127,
                                },
                                false,
                              ]
                            }
                            testID="PaymentCardSmallTestID"
                          >
                            <View
                              style={
                                Array [
                                  Object {
                                    "flexDirection": "row",
                                    "justifyContent": "space-between",
                                  },
                                  Object {
                                    "alignItems": "center",
                                  },
                                ]
                              }
                            >
                              <RNSVGSvgView
                                align="xMidYMid"
                                bbHeight={24}
                                bbWidth={24}
                                focusable={false}
                                height={24}
                                meetOrSlice={0}
                                minX={0}
                                minY={0}
                                style={
                                  Array [
                                    Object {
                                      "backgroundColor": "transparent",
                                      "borderWidth": 0,
                                    },
                                    Object {
                                      "flex": 0,
                                      "height": 24,
                                      "width": 24,
                                    },
                                  ]
                                }
                                vbHeight={24}
                                vbWidth={24}
                                width={24}
                              >
                                <RNSVGGroup
                                  fill={
                                    Object {
                                      "payload": 4278190080,
                                      "type": 0,
                                    }
                                  }
                                >
                                  <RNSVGPath
                                    d="M8.75415 6.58813h6.48917v11.662H8.75415z"
                                    fill={
                                      Object {
                                        "payload": 4285952960,
                                        "type": 0,
                                      }
                                    }
                                    propList={
                                      Array [
                                        "fill",
                                      ]
                                    }
                                  />
                                  <RNSVGPath
                                    d="M9.16616 12.4181c-.0018-2.2753 1.04274-4.42514 2.83254-5.82997-3.04065-2.38965-7.4072-2.04157-10.03091.79961-2.62372 2.84116-2.62372 7.22156 0 10.06276 2.62371 2.8412 6.99026 3.1892 10.03091.7996-1.7904-1.4053-2.83497-3.556-2.83254-5.832Z"
                                    fill={
                                      Object {
                                        "payload": 4293591067,
                                        "type": 0,
                                      }
                                    }
                                    propList={
                                      Array [
                                        "fill",
                                      ]
                                    }
                                  />
                                  <RNSVGPath
                                    d="M23.2411 17.4813v-.4132h.1534v-.0841h-.3908v.0841h.1535v.4132h.0839Zm.7588 0v-.4981h-.1198l-.1378.3426-.1378-.3426h-.1199v.4981h.0846v-.3758l.1292.324h.0877l.1293-.3248v.3766h.0845ZM23.9986 12.4181c-.0001 2.8395-1.6214 5.4296-4.1754 6.6705-2.554 1.2408-5.5922.9144-7.8244-.8405 1.7891-1.406 2.8336-3.5555 2.8336-5.831 0-2.2754-1.0445-4.42495-2.8336-5.83097 2.2322-1.75494 5.2704-2.0813 7.8244-.84048 2.554 1.24082 4.1753 3.83098 4.1754 6.67045v.002Z"
                                    fill={
                                      Object {
                                        "payload": 4278231519,
                                        "type": 0,
                                      }
                                    }
                                    propList={
                                      Array [
                                        "fill",
                                      ]
                                    }
                                  />
                                </RNSVGGroup>
                              </RNSVGSvgView>
                            </View>
                            <View
                              style={
                                Object {
                                  "height": 8,
                                }
                              }
                            />
                            <Text
                              allowFontScaling={false}
                              color="grey-700"
                              defaultColor="black"
                              defaultWeight="Regular"
                              ellipsizeMode="tail"
                              font="TitilliumWeb"
                              fontStyle={
                                Object {
                                  "fontSize": 12,
                                }
                              }
                              numberOfLines={1}
                              style={
                                Array [
                                  Object {
                                    "fontSize": 12,
                                  },
                                  Object {
                                    "color": "#555C70",
                                    "fontFamily": "Titillium Web",
                                    "fontStyle": "normal",
                                    "fontWeight": "400",
                                  },
                                ]
                              }
                              weight="Regular"
                            >
                              •••• 9900
                            </Text>
                          </View>
                        </View>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
